textcursor.go

```diff
@@ -0,0 +1,157 @@
+// Copyright 2015 The Go Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package gxui
+
+import "github.com/google/gxui/interval"
+
+// TextCursor represents a text position and selection. A selection is defined as
+// the pair of cursor indices sorted least to greatest. When transforming a
+// cursor with no selection, mark should equal index.
+type TextCursor struct {
+	index int
+	mark  int
+}
+
+// NewTextCursor returns a new TextCursor spanning the given indices. If mark equals
+// index, the cursor will have no selection.
+func NewTextCursor(index int, mark int) TextCursor {
+	return TextCursor{index, mark}
+}
+
+// Index returns the cursor index.
+func (c TextCursor) Index() int {
+	return c.index
+}
+
+// Mark returns the cursor mark index.
+func (c TextCursor) Mark() int {
+	return c.mark
+}
+
+// Range returns the difference between indices ordered highest and lowest.
+func (c TextCursor) Range() int {
+	if c.index > c.mark {
+		return c.index - c.mark
+	}
+	return c.mark - c.index
+}
+
+// Selection returns the cursor indices ordered least to greatest.
+func (c TextCursor) Selection() (int, int) {
+	if c.index < c.mark {
+		return c.index, c.mark
+	}
+	return c.mark, c.index
+}
+
+// HasSelection returns whether this cursor spans a range greater than zero.
+func (c TextCursor) HasSelection() bool {
+	return c.Range() > 0
+}
+
+// Invert returns the cursor in opposite order.
+func (c TextCursor) Invert() TextCursor {
+	return TextCursor{c.mark, c.index}
+}
+
+// Offset returns the cursor with indices + i.
+func (c TextCursor) Offset(i int) TextCursor {
+	return TextCursor{c.index + i, c.mark + i}
+}
+
+// NewSelection returns a cursor oriented as c with given selection.
+// If start is greater than end, the arguments will be reversed.
+func (c TextCursor) NewSelection(start, end int) TextCursor {
+	if start > end {
+		start, end = end, start
+	}
+	if c.index < c.mark {
+		return TextCursor{start, end}
+	}
+	return TextCursor{end, start}
+}
+
+// IsAtStart returns true if the cursor index is at the start of a selection.
```

Comment:
this is an unfortunate carry over from the old code. Ideally I'd like to completely remove this method.